torch/optim/adagrad.py

```diff
@@ -63,6 +64,7 @@ def step(self, closure=None):
                 clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])
 
                 if p.grad.data.is_sparse:
+                    grad.contiguous()  # the update is non-linear so indices must be unique
```

Comment:
the goal is for sparse tensors to be usable by anyone. I agree that it would be nicer to have `contiguous` (out of place) and `contiguous_` (in place)